### PR TITLE
[Merged by Bors] - adding `Functor.sum'`, dual to `Functor.prod'`.

### PR DESCRIPTION
**Pull request by sinhp**
We add `Functor.sum'` a variant of `Functor.sum`. This situation is dual to [docs#CategoryTheory.Functor.prod'](https://leanprover-community.github.io/mathlib4_docs/find/?pattern=CategoryTheory.Functor.prod%27#doc). 

We also add `inl_sum'` and `inr_sum'` which prove that the sum functor `F.sum' G` precomposed with the left (resp. right) inclusion functor is isomorphic to `F` (resp. `G`).  

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by mo271 — APPROVED** [review]

File: Mathlib/CategoryTheory/Sums/Basic.lean [hunk 1]
```diff
@@ -187,6 +187,34 @@ def sum (F : A ⥤ B) (G : C ⥤ D) : Sum A C ⥤ Sum B D
...
 #align category_theory.functor.sum CategoryTheory.Functor.sum
 
+/-- Similar to `sum`, but both functors land in the same category `C` -/
+def sum' (F : A ⥤ C) (G : B ⥤ C) : Sum A B ⥤ C
+    where
+  obj X :=
+    match X with
+    | inl X => F.obj X
+    | inr X => G.obj X
+  map := @fun X Y f =>
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Sums/Basic.lean:197]: The following syntax is better. (Do not hesitate doing the same change in all the file.)
```suggestion
  map {X Y} f :=
```

File: Mathlib/CategoryTheory/Sums/Basic.lean [hunk 2]
```diff
@@ -187,6 +187,34 @@ def sum (F : A ⥤ B) (G : C ⥤ D) : Sum A C ⥤ Sum B D
...
+    | inr _, inr _, f => G.map f
+  map_id := @fun X => by cases X <;> erw [Functor.map_id]
+  map_comp := @fun X Y Z f g =>
+    match X, Y, Z, f, g with
+    | inl _, inl _, inl _, f, g => by erw [F.map_comp]
+    | inr _, inr _, inr _, f, g => by erw [G.map_comp]
+
+/-- The sum `F.sum' G` precomposed with the left inclusion functor is isomorphic to `F` -/
+@[simps!]
+def inl_sum' (F : A ⥤ C) (G : B ⥤ C) : Sum.inl_ _ _ ⋙ F.sum' G ≅ F :=
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Sums/Basic.lean:209]: There must be no `_` is "definitions" (contrary to lemmas/theorems).
```suggestion
def inlCompSum' (F : A ⥤ C) (G : B ⥤ C) : Sum.inl_ _ _ ⋙ F.sum' G ≅ F :=
```

File: Mathlib/CategoryTheory/Sums/Basic.lean (hunk 1 above)
Comment (sinhp) [reply Mathlib/CategoryTheory/Sums/Basic.lean]: Indeed, this syntax looks neater. I just changed all such cases in the file uniformly.

File: Mathlib/CategoryTheory/Sums/Basic.lean (hunk 2 above)
Comment (sinhp) [reply Mathlib/CategoryTheory/Sums/Basic.lean]: Still getting used to the new naming convention :)
I changed the name of the definitions and also removed `_` in both definitions `inlCompSum'` and `inrCompSum'`.

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**